shared/middleware/helpers/storage_apps.rb

```diff
@@ -248,4 +248,58 @@ def self.merged_row_value(row, channel_id:, is_owner:)
       }
     )
   end
+
+  #
+  # Given an encrypted channel id, attempt to determine the channel's
+  # project type.
+  # This isn't always possible - we aren't consistent about storing project
+  # type information with the channel.
+  #
+  # @param [String] channel_id - an encrypted channel id
+  # @return [String] The discovered project type, or 'unknown' if the type
+  #   can't be determiend with given information.
+  # @raise [NotFound] if the channel does not exist or is not shareable.
+  #
+  def project_type_from_channel_id(channel_id)
+    project_type_from_merged_row(get(channel_id))
+  end
+
+  private
+
+  #
+  # Discovering a channel's project type is a real mess.  We don't usually
+  # need to do this because the project type is usually part of the URL,
+  # but for a few APIs this is needed.
+  #
+  # @param [Hash] row - A storage_apps merged row value as returned by
+  #   `merged_row_value` or `get`.
+  # @returns [String] The discovered project type, or 'unknown' if the type
+  #   can't be determined with given information.
+  #
+  def project_type_from_merged_row(row)
+    # We can derive channel project type from a few places.
+    #
+    # 1. The `project_type` column in the storage_apps table.
+    #    This is often NULL for "hidden" levels, which includes project-backed
+    #    script levels.
+    return row[:projectType] if row[:projectType]
+
+    # 2. The projectType property in the `value` JSON column.
+    #    Apparently this is sometimes filled out _instead_ of the column.
+    return row['projectType'] if row['projectType']
```

Comment:
mostly harmless, but no need to check this here -- `JSON.parse(row[:value])['projectType']` is only ever set from the database row, so if this is set then you'll always have returned in step 1 above.

Reply:
That's apparently not true. I haven't found the production code path that does this, but [our tests do it](https://github.com/code-dot-org/code-dot-org/blob/eb8c70c29dd4bb4db611ae328f275764f0990be3/shared/test/middleware/helpers/test_storage_apps.rb#L17) and I located an example in real data, created yesterday, that has a level and projectType in its JSON blob but a NULL project_type column:

```
mysql> select id, created_at, project_type, value from storage_apps where id=<hidden>;
+----------+---------------------+--------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id       | created_at          | project_type | value                                                                                                                                                                                                                                                                                                                                                               |
+----------+---------------------+--------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| <hidden> | 2018-04-17 20:39:28 | NULL         | {"hidden":true,"createdAt":"2018-04-17T20:39:28.000+00:00","updatedAt":"2018-04-17T20:43:21.925+00:00","id":"<hidden>","isOwner":true,"publishedAt":null,"projectType":"gamelab","level":"/projects/gamelab","migratedToS3":true,"thumbnailUrl":"/v3/files/<hidden>/.metadata/thumbnail.png"} |
+----------+---------------------+--------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

`StorageApps.create` dumps an arbitrary hash into `row[:value]` so it's one possible path to end up with `JSON.parse(row[:value])['projectType']` without `row[:project_type]`.

https://github.com/code-dot-org/code-dot-org/blob/eb8c70c29dd4bb4db611ae328f275764f0990be3/shared/middleware/helpers/storage_apps.rb#L17-L30

But I can't trace back to exactly where this happens, and I'm wondering if we mess with the storage_apps table directly anywhere else.  I can keep looking.

Reply:
Seems to be the case for a lot of projects, actually:
```
mysql> select count(id) from storage_apps where id > 52387852 and project_type is null and value LIKE '%projectType%' and value not LIKE '%"projectType":null%';
+-----------+
| count(id) |
+-----------+
|    102165 |
+-----------+
1 row in set (0.73 sec)
```

Reply:
ahh, good catch! it is very sad that we are so all over the place on this. FWIW, Asher wrote a script to rectify some of this: https://github.com/code-dot-org/code-dot-org/blob/ca6ea8a8c63a96a932f25834a5ac08c687259bf8/bin/oneoff/backfill_data/storage_apps_project_type_and_standalone so we may be closer than we think to fixing it.

Reply:
Good find on that script - I'll come back to that if we have time (after May) to do some cleanup here.